README.md
# BitBeast
A game by Cheese and Bacon Games
http://cheeseandbacon.org/

This is the source code for Bit Beast.

Bit World is a digital realm full of digital eggs. No one knows how this world came into existence, or where the eggs come from.
However, what we do know is that using an Android device, coupled with this software, you can acquire one of these eggs.

Once on your device, the egg will begin to mature and eventually hatch into a mysterious creature, known as a Bit Beast!

Raise your Bit Beast well, pit it against other Bit Beasts in battle, and become an expert Bit Beast handler!

Bit Beast is a virtual pet inspired by the great toys from the 90s, such as Tamagotchi and Digimon. You raise and care for your Bit Beast by feeding him, cleaning him, giving him medicine when he is sick, playing with him, and so on.

As your Bit Beast grows up, he will evolve into different forms. There are several different evolutionary branches that your Bit Beast can move along, and what he becomes is dependent on how well you raise him. Pay special attention to what kind of food you feed him!

To interact with your Bit Beast, simply run the app. It runs like any other standard app and can be shut down whenever you want. When you next run the app, it calculates what happened during the intervening period and updates accordingly.

At any time, you can pause your Bit Beast, and nothing will happen to him until you unpause him later. You can also configure his bedtime so that he can be awake whenever you are!

Bit Beast makes use of several of Android's fun hardware features, if they are available. If a given feature is not available on your device, don't panic! You can still use the app! The feature will simply not be available when you play.

Here is how Bit Beast puts some of these features to good use:

- You can pet your Bit Beast (and interact with the rest of the UI) using the touchscreen if you have one. If not, you can use your trackball or dpad to navigate.

- You can tap the screen when on the main view to activate speech recognition, and give your Bit Beast commands, praise, or scolding.

- You can shake your device to earn bits (this one uses the accelerometer).

- You can turn on your GPS and earn bits based on either speed or distance traveled.

- You can connect to another Bit Beast on another device via Bluetooth and battle them!

View the complete [changelog](development/changelog.txt).

Here is an explanation of the permissions required to install the app, and why they are needed: [permissions](development/permissions.txt)

While the app is already complete, there will almost certainly be updated versions released in the future to fix bugs, tweak numbers, etc. If you find a bug, or something that just doesn't seem right, please let us know! You can contact us via email, or head over to the forums, using the links found below.

After you have played with your Bit Beast for a bit (pun intended!), please rate it on Google Play to let us, and everyone else, know what you think of the game!

Learn more about Bit Beast at: http://cheeseandbacon.org/bit-beast/

## License
This game's source code is licensed under the MIT License. See [LICENSE.txt](development/LICENSE.txt) for the full license text.

This game's data is not under the same license as the source code. See [CONTENT-LICENSE.txt](development/CONTENT-LICENSE.txt) for game data licensing details.


development/changelog.txt
1.1.0 - 2018/06/27
 - Two new stats added: Dexterity and Stamina
   These each function kind of like one half of strength
   Strength is attack damage as well as health
   Stamina provides extra health that must be removed before strength can be damaged, thus acting like a "shield"
   Dexterity provides a damage bonus but gets used up as you attack,
   thus giving you a large damage burst at the beginning of a battle
 - Training now increases max Dexterity and max Stamina in addition to max Strength
 - Each pet evolution branch now represents a distinct "class," with its own unique set of stat priorities
 - Added levels and experience. When you level up, you get stat points to spend on improving your stats
 - Split the store into 4 separate sections
 - Added "permanent" items which are one-time purchases kept in the yard
   Rumor has it they have other benefits...
 - Consumable items can now grant a wide variety of temporary buffs
 - Added some new consumable items
 - All consumable items now have a "primary affected stat," which is what limits their use
   I.e., most food items won't allow you to eat them once your food bar is full
   However, junk food won't allow you to eat it once your HAPPY bar is full
 - Weight now plays a larger role in things. Specifically, being obese (>30 kg) is penalized more
 - The old win/loss ratio is multiplayer win/loss, and there is now a new singleplayer win/loss ratio, too
   Note that old cemetery entries will look a little silly having "0" for so many new things
 - Rewards for various activities are now all merged into one system, in which a Rewards window pops up after
   you are done playing. This means that everything has experience and bit rewards
 - Bit reward on battle victory scaled by max stat difference of opponents
 - Max stat increase on successful training now scales by AI opponent difficulty
 - Max stat gains from training are now capped on a per-tick basis
 - Increased the rate of max stat gain from training
 - Natural temperature change slowed dramatically
 - Machine temperature change increased dramatically
 - Temperature machines now shut off if they go very far past the optimum temperature
 - Rewards for bricks are scaled based on number of bricks
 - Added some more brick layouts
 - Added powerups to the paddle games
 - Added new sound effects throughout
 - Improved paddle game AI and physics
 - Added more brick layouts to Bricks
 - A lot more things regenerate in real time now instead of every tick (15 minutes)
 - You can now watch your Bit Beast consume things that you give it
 - Changed a number of messages
   They are theoretically more fun than ever!
 - You can now view the status screen, clean the yard, and change the temperature setting even if your Bit Beast is dead
 - Fixed some issues with Bit Beast death
 - Bit Beasts are "cleaned" when they die
 - There is only one sound effect for thought bubbles now
 - Permanent items can now be dragged when Bit Beast is dead
 - Pets will now properly center on held items
 - Unusable buttons are now grayed out
 - Fixed egg and tombstone being solid colors in the widget
 - The widget will no longer create read error messages
 - The widget now updates once every hour
 - Added new information to the widget
 - Improved widget scaling
 - The widget now updates whenever pet data is saved
 - Changed optimum temperature from exactly 20 degrees C to 20 +/- 1
 - Reduced the age bonus for bits gained
 - Improved the variety of shadow opponents. They are also tougher now
 - Changed the way favorite food is stored in save data. Note that a Bit Beast from pre-1.1.0 will forget
   its favorite food because of this
 - Removed the screen orientation option
 - Updated the Help screens
 - Fixed numerous bugs where an Android configuration change would cause loss of data
 - Balance tweaks
 - Fixed bug with evolution branch choosing code
 - Fixed major bug with food categories
 - Fixed minor bug with food categories
 - Fixed a few instances where speech would not wake up the Bit Beast
 - Improved color picker dialog
 - Improved RNG
 - Fixed a bug where dialogs would display text in the wrong font
 - Fixed a variety of image-related bugs
 - Moved save data to internal storage
 - Made file saving atomic
 - Removed screenshot feature
 - Improved permission handling
 - Bit Beasts must be the same version to battle
 - Made tick processing more deterministic
 - Moved the project into Android Studio and updated build tools to latest versions
 - Updated build target to latest version of Android (API level 27)
 - Updated minimum API level to 21
 - Minor code cleanup and improvements
 - Minor graphical improvements
 - And more!

1.0.5 - 2012/6/1
 - Tweaked weight loss from training and battles... again
 - The Bit Beast's shadow is now randomly chosen from all forms of the Bit Beast at its age

1.0.4 - 2012/5/29
 - Tweaked weight loss from training and battles again
 - Improved the way temperature changes
 - Tweaked temperature change rates

1.0.3 - 2012/5/27
 - Fixed an OutOfMemoryError (hopefully)
 - Reduced filesize and memory usage slightly
 - Increased weight loss from training and battling, and made it somewhat fuzzy
 - Slight improvement to the RNG
 - Tweaked some brick layouts
 - Accelerometer and GPS minigames now cause weight loss
 - Made shadowboxing opponent tougher
 - Bit gains now scale based on age tier

1.0.2 - 2012/5/24
 - Fixed a pixel conversion function
 - Some minor tweaks

1.0.1 - 2012/5/24
 - Graphical improvements

1.0.0 - 2012/5/22
 - Initial release version!


src/org/cheeseandbacon/bitbeast/Sound.java
package org.cheeseandbacon.bitbeast;

public class Sound{
	static final int IGNORE=0;
	static final int BEGIN=1;
	static final int AC=BEGIN;
	static final int BATHE=2;
	static final int CLEAN=3;
	static final int DRINK=4;
	static final int EAT=5;
	static final int EVOLUTION_1=6;
	static final int EVOLUTION_2=7;
	static final int GAME_HIT_BRICK=8;
	static final int GAME_HIT_PADDLE=9;
	static final int GAME_HIT_WALL=10;
	static final int GAME_LOSS=11;
	static final int GAME_WIN=12;
	static final int HEATER=13;
	static final int MEDICINE=14;
	static final int POOP=15;
	static final int THOUGHT=16;
	static final int GAME_RESET_LOSS=17;
	static final int TOGGLE_LIGHT=18;
	static final int DIE=19;
	static final int SPEECH_START=20;
	static final int SPEECH_STOP=21;
	static final int BATTLE_HIT=22;
	static final int BATTLE_MISS=23;
	static final int PERMA_ITEM_GRAB=24;
	static final int PERMA_ITEM_DROP=25;
	static final int LEVEL_UP=26;
	static final int SPEND_STAT_POINT=27;
	static final int NO_STAT_POINTS=28;
	static final int GAME_DRAW=29;
	static final int GAME_RESET_WIN=30;
	static final int POWERUP_GET=31;
	static final int POWERUP_SPAWN=32;
	static final int ITEM_SOLD=33;
	static final int EQUIPPED=34;
	static final int UNEQUIPPED=35;
	static final int ITEM_SOLD_ALL=36;
	static final int END=37;
}
